File: src/bun/bun_serve/src/routes/cache.ts
import { cacheService } from '../services/cache.ts';
import type { CacheResponse } from '../types.ts';

// The TTL is part of the response contract and must match what is written
// to Redis. See contracts/rest/canonical-payloads.md.
const CACHE_TTL_SECONDS = 300;

export async function cacheHandler(request: Request): Promise<Response> {
  try {
    const url = new URL(request.url);
    const key = url.searchParams.get('key') || 'test';
    const cached = await cacheService.get(key);

    if (cached !== null) {
      const response: CacheResponse = {
        key,
        value: cached,
        cached: true,
        ttl: CACHE_TTL_SECONDS,
      };

      return new Response(JSON.stringify(response), {
        status: 200,
        headers: {
          'Content-Type': 'application/json',
        },
      });
    }

    const value = `cached_data_${key}_${Date.now()}`;
    await cacheService.set(key, value, CACHE_TTL_SECONDS);

    const response: CacheResponse = {
      key,
      value,
      cached: false,
      ttl: CACHE_TTL_SECONDS,
    };

    return new Response(JSON.stringify(response), {
      status: 200,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  } catch (error) {
    return new Response(JSON.stringify({
      error: 'Internal Server Error',
      message: 'An error occurred while processing the request',
    }), {
      status: 500,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }
}

File: src/python/django/benchmark/views/cache.py
from rest_framework.decorators import api_view
from rest_framework.response import Response
from benchmark.services.cache_service import CacheService
from datetime import datetime

# The TTL is part of the response contract and must match what is written
# to Redis. See contracts/rest/canonical-payloads.md.
CACHE_TTL_SECONDS = 300

@api_view(['GET'])
def cache(request):
    key = request.GET.get('key', 'test')
    value, was_cached = CacheService.get_or_set(key)

    return Response({
        'key': key,
        'value': value,
        'cached': was_cached,
        'ttl': CACHE_TTL_SECONDS,
        'timestamp': datetime.utcnow().isoformat()
    })
